fdbserver/ConsistencyScan.actor.cpp

```diff
@@ -0,0 +1,1090 @@
+		state std::map<UID, StorageServerInterface> tssMapping;
+		bool keyServerResult = wait(getKeyServers(self->db, keyServerPromise, keyServersKeys));
+		if (keyServerResult) {
+			state std::vector<std::pair<KeyRange, std::vector<StorageServerInterface>>> keyServers =
+			    keyServerPromise.getFuture().get();
+
+			// Get the locations of all the shards in the database
+			state Promise<Standalone<VectorRef<KeyValueRef>>> keyLocationPromise;
+			bool keyLocationResult = wait(getKeyLocations(self->db, keyServers, keyLocationPromise, false));
+			if (keyLocationResult) {
+				state Standalone<VectorRef<KeyValueRef>> keyLocations = keyLocationPromise.getFuture().get();
+
+				// Check that each shard has the same data on all storage servers that it resides on
+				wait(::success(checkDataConsistency(self->db,
+				                                    keyLocations,
+				                                    self->configuration,
+				                                    tssMapping,
+				                                    false /* quiescentCheck */,
+				                                    false /* tssCheck */,
+				                                    true /* firstClient */,
+				                                    false /* failureIsError */,
+				                                    0 /* clientId */,
+				                                    1 /* clientCount */,
+				                                    false /* distributed */,
+				                                    false /* shuffleShards */,
+				                                    1 /* shardSampleFactor */,
+				                                    deterministicRandom()->randomInt64(0, 10000000),
+				                                    self->finishedRounds /* repetitions */,
+				                                    &(self->bytesReadInPrevRound),
+				                                    self->restart,
+				                                    self->maxRate,
+				                                    self->targetInterval,
+				                                    self->progressKey)));
+			}
+		}
+	} catch (Error& e) {
+		if (e.code() == error_code_transaction_too_old || e.code() == error_code_future_version ||
+		    e.code() == error_code_wrong_shard_server || e.code() == error_code_all_alternatives_failed ||
+		    e.code() == error_code_process_behind || e.code() == error_code_actor_cancelled)
+			TraceEvent("ConsistencyScan_Retry").error(e); // FIXME: consistency check does not retry in this case
+		else
+			throw;
+	}
+
+	TraceEvent("ConsistencyScan_FinishedCheck");
+
+	//  Update the ConsistencyScanInfo object and persist to the database
+	csInfo.last_round_finish = StorageMetadataType::currentTime();
+	csInfo.finished_rounds = self->finishedRounds + 1;
+	auto duration = csInfo.last_round_finish - csInfo.last_round_start;
+	csInfo.smoothed_round_duration.setTotal((double)duration);
+	csInfo.progress_key = self->progressKey;
```

Comment:
This seems to be only done upon error or finishing a complete round, which means progress may not be saved for a very long time or at all.  Is that correct?

Reply:
Not the entire round. It's updated after validating a single shard across all storage server replicas.

Reply:
I only see one place where a `ConsistencyScanInfo` object's `progress_key` member is updated and saved to the database, which is right here, and this update increments `csInfo.finished_rounds` and sets `csInfo.last_round_finished` as well which is consistent with the idea that this is after an entire round has finished.

Reply:
Good catch. I realized that I was only updating the in-memory progressKey after each shard validation and not the on-disk object. I have made the change to persist the progressKey after each shard is verified.